Fix and test compliance of TLS 1.3 record protection

hanno-arm opened the pull request:
__Context:__ This PR continues the upstreaming of the TLS 1.3 prototype from https://github.com/hannestschofenig/mbedtls/tree/tls13-prototype.

__What:__
* Fix a non-compliance bug in the TLS 1.3 record protection mechanisms previously upstreamed in https://github.com/ARMmbed/mbedtls/pull/3315: The computation of the AEAD Additional Data (AAD) was wrong, as indicated in the respective commit message.
* Add known answer unit tests for TLS 1.3 record protection mechanisms that would have caught the bug, based on https://tools.ietf.org/html/rfc8448 and https://tls13.ulfheim.net/.
* As part of providing those unit tests, upstream the TLS 1.3 specific function `mbedtls_ssl_tls13_populate_transform()`, which is the TLS 1.3 sibling of the existing `ssl_populate_transform()` and establishes an SSL transform structure from raw key material and connection metadata.

hanno-arm commented:
@paul-elliott-arm @bensze01 When do you think you'll find time to review this?

bensze01 commented on tests/suites/test_suite_ssl.function:
```diff
@@ -3807,6 +3807,92 @@ MBEDTLS_SSL_TLS1_3_LABEL_LIST
+    ASSERT_ALLOC( buf, ciphertext->len );
+    rec.type   = MBEDTLS_SSL_MSG_APPLICATION_DATA;
+    mbedtls_ssl_write_version( MBEDTLS_SSL_MAJOR_VERSION_3,
+                               MBEDTLS_SSL_MINOR_VERSION_3,
```
Shouldn't this be `MINOR_VERSION_4`?

hanno-arm replied:
No, it's deliberate that this is the legacy TLS 1.2 version identifier, to avoid middlebox interference. You cannot tell TLS 1.2 and TLS 1.3 records apart on the wire. I'll add a comment.

hanno-arm commented on library/ssl_tls13_keys.c:
```diff
@@ -346,4 +347,112 @@ int mbedtls_ssl_tls1_3_evolve_secret(
+    (void) ssl;
+#endif
+
+    ciphersuite_info = mbedtls_ssl_ciphersuite_from_id( ciphersuite );
```
Generally, I think this is an assertion rather than a genuine check, so we might want to omit it (we certainly can't include assertions for each and everything, and it's up to us to determine where to draw the line) but seeing that the error code returned from the TLS 1.2 version is not `MBEDTLS_ERR_SSL_INTERNAL_ERROR` makes me question that impression. In any case, it should be the same for uniformity, as you say. I'll change it.

hanno-arm replied:
Done.

hanno-arm commented on library/ssl_tls13_keys.c:
```diff
@@ -346,4 +347,112 @@ int mbedtls_ssl_tls1_3_evolve_secret(
+    transform->ivlen       = traffic_keys->iv_len;
+    transform->maclen      = 0;
+    transform->fixed_ivlen = transform->ivlen;
+    transform->minlen      = transform->taglen + 1;
```
Yes, this is because of the true record content type which is added to the plaintext in TLS 1.3.

However, I realize that the `+1` should actually be a `+ padding_granularity`, and I've fixed and documented the code accordingly. 

(NB: I don't know if there's any relevance left to `transform->minlen` TBH)

hanno-arm commented:
Thank you very much @bensze01 for your review! I believe to have addressed your comments. Could you please take another look?

bensze01 commented on library/ssl_tls13_keys.c:
```diff
@@ -367,12 +367,19 @@ int mbedtls_ssl_tls13_populate_transform( mbedtls_ssl_transform *transform,
-        return( MBEDTLS_ERR_SSL_INTERNAL_ERROR );
+        MBEDTLS_SSL_DEBUG_MSG( 1, ( "cipher info for %u not found",
+                                    ciphersuite_info->cipher ) );
+        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA )
```
```suggestion
        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
```
Missing semicolon.

hanno-arm replied:
Oh, embarrassing. Fixed.

bensze01 commented:
Also, what do you think about removing the section guarded by `MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL` from `ssl_tls12_populate_transform()`, as I mentioned previously?

hanno-arm commented:
@bensze01 Good point. I have removed the code in question and added an assertion to `ssl_tls12_populate_transform()` to double-check that we're not using it for TLS 1.3.

bensze01 reviewed (dismissed):
Thanks for the changes, Hanno. There are a couple of merge conflicts that developed since last week, but aside from that, this seems OK to me.

hanno-arm commented:
@bensze01 I have rebased this on top of `development`, resolving a trivial merge conflict stemming from the removal of compression support for 3.0.

bensze01 reviewed (approved):
Thanks for the rebase, looks good to me.

hanno-arm commented:
@paul-elliott-arm Do you think you have time to review this PR?

gabor-mezei-arm reviewed (dismissed):
LGTM

hanno-arm commented:
@bensze01 @gabor-mezei-arm I rebase this on top of `development`, resolving trivial merge conflicts. Could you please re-review?

hanno-arm commented:
Something went wrong. Needs work.

hanno-arm commented:
The merge of CID and TLS 1.3 padding granularity changed the default padding granularity for TLS 1.3 from 1 to 16. However, the unit tests added in this PR were specific to granularity 1. I have added some commits to address that.

bensze01 reviewed (approved):
Looks good to me.

gabor-mezei-arm reviewed (approved):
Looks good to me.
